src/hotspot/share/opto/addnode.cpp

```diff
@@ -920,9 +920,24 @@ const Type* XorINode::Value(PhaseGVN* phase) const {
   if (in1->eqv_uncast(in2)) {
     return add_id();
   }
+  // result of xor can only have bits sets where any of the
+  // inputs have bits set. lo can always become 0.
+  const TypeInt* t1i = t1->is_int();
+  const TypeInt* t2i = t2->is_int();
+  if ((t1i->_lo >= 0)        &&
+      (t1i->_hi > 0)         &&
+      (t1i->_hi <= max_power_of_2<jint>()) &&
```

Comment:
I think you could use `<= std::numeric_limits<T>::max()` as the upper bound condition, and use `(jint)(next_power_of_2((uint)t1i->_hi) - 1)` to produce the mask in an overflow-conscious way.

Reply:
max_power_of_2<jint>() is less than std::numeric_limits<T>::max(). 

But I noticed I got the bounds wrong - is should be strictly less:
t1i->_hi < max_power_of_2<jint>()

Reply:
Right, that's why I widened to an `uint` (since `next_power_of_2((uint)jint::max())` is well-defined) then cast back to a `jint` after subtracting 1.

Reply:
Another way of expressing the same in an overflow-conscious way (without type conversion) is `round_down_power_of_2(t1i->_hi) + round_down_power_of_2(t1i->_hi) - 1`

Reply:
I agree - that is a much better solution! Thanks Claes!